### PR TITLE
[vulkan] Add integer dot product (4xint8, 4xuint8) tensorization for the vulkan SPIR-V target.

### PR DESCRIPTION
**Pull request by mei-ye**
@masahi

[vulkan] Add integer dot product (4xint8, 4xuint8) tensorization for the vulkan SPIR-V target. Currently only autotvm path is supported.
    
Prerequisites for compilation: (1) Use VulkanSDK 1.2.198 release with SPIR-V integer dot product suppport  (2) set(USE_SPIRV_KHR_INTEGER_DOT_PRODUCT ON) in config.cmake and build (3) Use a driver that supports VK_KHR_shader_integer_dot_product extension.
    
The compiled binary can only be run on a hardware that supports relevant ISA. This work is tested on AMD RDNA2 famillies (e.g., Rembrandt and RX6800).
    
To compile on a device that supports this extension, use target: vulkan -from_device=0
    
To compile on a device that supports int8 but does not support this extension, add "-supports_integer_dot_product=1" or "-mattr=+dotprod" to the target string.
    
To support pre-released vulkan and SPIR-V extensions, we need SPIR-V tool and header file from Khronos github, use the option: USE_KHRONOS_SPIRV in config.cmake.
    
Example to use this feature can be found in: gallery/how_to/deploy_models/deploy_prequantized.py and gallery/how_to/deploy_models/deploy_prequantized_tflite.py

### CODE REVIEW

**Comment by masahi** [comment]
@mei-ye Can you clean up the commit history? (reset to `main` and apply [7eea90b](https://github.com/apache/tvm/pull/10391/commits/7eea90bae0031d0dd6db668314f0013e95fa6f76))

**Comment by masahi** [comment]
Please remove the change in `deploy_prequantized.py` and `deploy_prequantized_tflite.py`. They run on CI and we don't intend to run e2e models on vulkan target there.

File: python/tvm/relay/op/strategy/cuda.py [hunk 1]
```diff
@@ -372,7 +376,7 @@ def conv2d_strategy_cuda(attrs, inputs, out_type, target):
             ic_chunk = in_channels // 4
 
             if (
-                target.kind.name == "cuda"
+                (target.kind.name == "cuda" or target.kind.name == "vulkan")
```
Comment (masahi) [inline python/tvm/relay/op/strategy/cuda.py:379]: `target.kind.name in ["cuda", "vulkan"]`. Three places.

File: python/tvm/topi/cuda/conv2d_alter_op.py [hunk 2]
```diff
@@ -83,14 +83,19 @@ def _alter_conv2d_layout(attrs, inputs, tinfos, out_type):
     cfg = dispatch_ctx.query(target, workload)
     if cfg.is_fallback:  # if is fallback, clear query cache and return None
         autotvm.task.clear_fallback_cache(target, workload)
-        return None
+        if (
+            "vulkan" not in target.keys
+            or "+dotprod" not in target.mattr
+            or not target.supports_integer_dot_product
+        ):
+            return None
```
Comment (masahi) [inline python/tvm/topi/cuda/conv2d_alter_op.py:91]: What is this change for?

**Comment by masahi** [comment]
@mei-ye Please make 

```
    if (
        "vulkan" not in target.keys
        or "+dotprod" in target.mattr
        or target.supports_integer_dot_product
    ):
```

more readable. For now this should only be True for cuda and vk with dot product support. But later I'll add opencl and rocm support too.

File: python/tvm/topi/cuda/conv2d_alter_op.py (hunk 2 above)
Comment (mei-ye) [reply python/tvm/topi/cuda/conv2d_alter_op.py:91]: If it is fall back and if dot product is to be used, make the flow to proceed to do the NCHW => NCHW4c layout transformation.

**Comment by masahi** [comment]
Ok @mei-ye the CI is green now. The changes I made are:

* Remove the `storage_rewrite.cc` change that was originally in @mei-ye's commit, but it turned out it is no longer necessary after @Lunderberg's PR https://github.com/apache/tvm/pull/9727 was merged. 
* Check tensorization condition in `batch_matmul` https://github.com/apache/tvm/pull/10391/commits/aa7328f09a8f0f30d4fbb1cfb8d0455748bb9547
* Do not run vk tests on int8 `batch_matmul` and `conv2d`. In particular, for some reason running int8 `batch_matmul` test on vk, disabled in https://github.com/apache/tvm/pull/10391/commits/839d2338a3a750278d77ece3de57c94559015420, was apparently the cause of the strange CI error problem.

**Review by masahi — APPROVED** [review]

**Pull request merged**